src/framework/mlt_chain.c

```diff
@@ -0,0 +1,356 @@
+/**
+ * \file mlt_chain.c
+ * \brief link service class
+ * \see mlt_chain_s
+ *
+ * Copyright (C) 2020 Meltytech, LLC
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ */
+
+#include "mlt_chain.h"
+#include "mlt_frame.h"
+#include "mlt_log.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+
+/** \brief private service definition */
+
+typedef struct
+{
+	int link_count;
+	int link_size;
+	mlt_link* links;
+	mlt_producer source;
+	mlt_profile source_profile;
+	mlt_producer begin;
+}
+mlt_chain_base;
+
+/* Forward references to static methods.
+*/
+
+static int producer_get_frame( mlt_producer parent, mlt_frame_ptr frame, int index );
+static void relink_chain( mlt_chain self );
+
+/** Construct a link.
+ *
+ * Sets the resource property to "<link>", the mlt_type to "mlt_producer",
+ * and mlt_service to "link".
+ *
+ * \public \memberof mlt_chain_s
+ * \return the new link
+ */
+
+mlt_chain mlt_chain_init( mlt_profile profile )
+{
+	mlt_chain self = calloc( 1, sizeof( struct mlt_chain_s ) );
+	if ( self != NULL )
+	{
```

Comment:
With regards to style, I am OK to make all new code not have spaces inside parentheses and opening braces on the same line as a statement. Also, I tend to not compare with 0 or NULL. For example,
```
mlt_chain self = calloc(1, sizeof(struct mlt_chain_s));
if (self) {
```
The choice is up to you; I am simply stating that I am trying to use a more normal style now because this old, unique style kind of bothers.

Reply:
Would you have any interest in a Make target or a script to run [Uncrustify ](https://github.com/uncrustify/uncrustify) to enforce a coding style? I have some experience with uncrustify.

Options would be:
1) "Fix" all source files
2) "Check" all source files and report an error if any do not comply

The check could be added to the weekly regression test (optional)

The fix could be used for:
1)  First "big bang" change to update all source files to match style
2) A quick run for developers before submitting code
3) A quick cleanup if the automated regression test reports non-compliant files (if we went that route) or occasional runs to catch up code that has deviated from the style

I think that we would not want to add it to the regression test right away.

Reply:
I am aware of code formatters and used `astyle` in Kino. I am not inclined to enforce it but perhaps willing to accept a mass conversion.